### ISSUE TITLE
IOS: Improve error reporting when biometry factor can't be used

### ISSUE THREAD

**Issue by hvge** [wultra/powerauth-mobile-sdk#58]
There are two separate minor issues related to biometry in our current IOS implementation:

### 1. TouchIDNotAvailable is not reported

`PA2ErrorCodeTouchIDNotAvailable` is returned only from `addBiometryFactor:callback:` method. The error code should be returned also when the signature is calculated and the biometry key is not available at the time. This may happen, if user has an already paired application with the biometry key stored in the keychain and then turns off TouchID on the device.

This topic needs a little bit more investigation. We should check an IOS keychain's behavior:
- Does IOS remove an application's record when the TouchID is turned off on global scale?
- If the record is still in the keychain, will this be valid, when user turns the feature on again?
- Should we remove the record automatically in this situations?

### 2. Signature calculation improvements

A private `signatureKeysForAuthentication:userCancelled:` & `biometryRelatedKeyUserCancelled:prompt:` methods should return a `NSError` instead of boolean about cancellation. There are also several other possible improvements:

1. Method declares that `userCancelled` pointer is nullable, but doesn't check whether it's null when writes the result. Despite the fact that the method is private, the implementation should check the nullability. This will be irrelevant once we change the interface and the error will be returned (of course, we should check, whether the pointer to error is valid)
2. Method should not generate a fake key, when the biometry is not available and should return an error in this case. Be aware that the fake key **must** be returned when auth by TouchID fails. We are using this to simulate an authentication attempt, like when user types a wrong PIN. The fake key generation should be moved to `biometryRelatedKeyUserCancelled:prompt:`

**Comment by petrdvorak** [comment]
Couple comments:

1) Generally, SDK user should check for biometry availability before trying to use it:

https://github.com/lime-company/powerauth-mobile-sdk/wiki/PowerAuth-SDK-for-iOS#check-touch-id-status

> Does IOS remove an application's record when the TouchID is turned off on global scale?

No, last time I checked. We should retest.

> If the record is still in the keychain, will this be valid, when user turns the feature on again?

Yes, last time I checked. We should retest.

(I did testing of these scenarios when writing the original SDK and this is why we have the three level of Touch ID / Face ID checks - system level, activation level, application level)

> Should we remove the record automatically in this situations?

No. We should keep the keys where they are, they are locked in the Secure Enclave and not accessible.

2) While I admit reducing this issue to `boolean` value is a bit too much, we should definitely provide some abstraction to the error handling, not just passing the low level Keychain error code. We could, however, include the original Keychain error in `userInfo` dictionary...

If we generate the fake key or not should depend on the low level Keychain error, that we decide to report as "fake successful authentication" (such as a situation with wrong finger on scanner, etc.).
